File: LeetCode/C/75-two-pointers-颜色分类.cpp
/*
 * @lc app=leetcode.cn id=75 lang=cpp
 *
 * [75] 颜色分类
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    void sortColors(vector<int> &nums) {
        int l = 0, r = nums.size() - 1;
        for (int i = 0; i <= r; i++) {
            if (nums[i] == 0)
                swap(nums[l++], nums[i]);
            if (nums[i] == 2)
                swap(nums[i--], nums[r--]);
        }
    }
};
// @lc code=end

File: LeetCode/Go/75-two-pointers-颜色分类.go
/*
 * @lc app=leetcode.cn id=75 lang=golang
 *
 * [75] 颜色分类
 */
package leetcode

// @lc code=start
func sortColors(nums []int) {
	var p0, p2 = 0, len(nums) - 1
	var cur = 0
	for cur <= p2 {
		if nums[cur] == 0 {
			nums[p0], nums[cur] = nums[cur], nums[p0]
			p0, cur = p0+1, cur+1
		} else if nums[cur] == 2 {
			nums[cur], nums[p2] = nums[p2], nums[cur]
			p2--
		} else {
			cur++
		}
	}
}

// @lc code=end
